src/runtime-rs/crates/hypervisor/src/qemu/qmp.rs

```diff
@@ -467,8 +473,111 @@ impl Qmp {
 
         Ok(())
     }
+
+    pub fn get_device_by_qdev_id(&mut self, qdev_id: &str) -> Result<PciPath> {
+        let format_str = |vec: &Vec<i64>| -> String {
+            vec.iter()
+                .map(|num| format!("{:02x}", num))
+                .collect::<Vec<String>>()
+                .join("/")
+        };
+
+        let mut path = vec![];
+        let pci = self.qmp.execute(&qapi_qmp::query_pci {})?;
+        for pci_info in pci.iter() {
+            if let Some(_device) = get_pci_path_by_qdev_id(&pci_info.devices, qdev_id, &mut path) {
+                let pci_path = format_str(&path);
+                return PciPath::try_from(pci_path.as_str());
+            }
+        }
+
+        Err(anyhow!("no target device found"))
+    }
+
+    pub fn hotplug_vfio_device(&mut self, vfiodev: &mut VfioDevice) -> Result<()> {
+        // FIXME: the first one might not the true device we want to passthrough.
+        // TODO multifunction=on case
+        let primary_device = vfiodev.devices.first_mut().unwrap();
+        info!(
+            sl!(),
+            "qmp hotplug vfio primary_device {:?}", &primary_device
+        );
+
+        let mut vfio_args = Dictionary::new();
+        let bdf = if !primary_device.bus_slot_func.clone().starts_with("0000") {
+            format!("0000:{}", primary_device.bus_slot_func.clone())
+        } else {
+            primary_device.bus_slot_func.clone()
+        };
+        vfio_args.insert("addr".to_owned(), "0x0".into());
+        vfio_args.insert("host".to_owned(), bdf.into());
+        vfio_args.insert("multifunction".to_owned(), "off".into());
+
+        let vfio_dev_add = VfioDeviceAdd {
```

Comment:
Is there any reason to use `struct VfioDeviceAdd` here? It looks like just an intermediate structure which does not do anything particularly.

Reply:
Good catch, I think we can just use the expressions used to initialise `VfioDeviceAdd`'s members here to set `qmp::device_add` members directly right in the next statement.  It would even make the code a bit shorter.

Reply:
Thx fixed